### DISCUSSION TITLE
Keyboard input broken in NixOS 24.05, but fixed in unreleased rofi version

### DISCUSSION THREAD

**Discussion by hallettj** [davatorium/rofi#2008] [Distro Specific]
I ran into a problem running Rofi in wayland (specifically in the Niri window manager) after upgrading to NixOS 24.05. Rofi stopped responding to any keyboard input in any mode, while apparently still intercepting key presses so I would have to switch to a virtual tty to kill Rofi. Everything was working fine in NixOS 23.11 which I think uses the same Rofi package version, `1.7.5+wayland3`

Thankfully the problem is fixed in recent commits! I tracked down the specific commit that fixed the problem to https://github.com/lbonn/rofi/commit/b04bedca44b34e276a75785bec2c42a83d7a79b7.

I'm submitting this report to suggest tagging a new release to get that fix into NixOS.

I have conferred with some other Niri users, and heard that the bug does not appear using the latest Arch package in Niri. So it seems to be some Nix-specific issue. I wish I knew more.

I have [log output](https://gist.github.com/hallettj/31da54d5e66ad9bdc9a21c9ee7f34e6d) - in particular when pressing keys I see this message in logs:

> (process:282581): libnkutils-bindings-CRITICAL **: 13:29:47.086: nk_bindings_seat_handle_key: assertion 'self->keymap != NULL' failed
[ 887915.163] wl_keyboard@12.key(652399, 9663496, 108, 0)

Also, thanks very much for the work on this project! Rofi is the only tool I have found anywhere that I can use to switch to open windows in my particular window manager.

**Comment by hallettj** [comment]
Oh I'm sorry, I think I got confused. I thought I was starting a discussion on the wayland fork, but I see this is the original rofi repo.

**Reply by lbonn** [reply]
@hallettj ah well I did not enable discussions in the fork :]

Hopefully we can release rofi mainline and wayland version, that would solve issues with these already fixed bugs.

**Reply by lbonn** [reply]
For the conditions in which this bug appears, maybe @samueldr knows more?

**Reply by samueldr** [reply]
The version in Nixpkgs is the last tagged version.

The fix (which was for that issue indeed) is more recent than `1.7.5+wayland3`.

 - https://github.com/lbonn/rofi/compare/1.7.5+wayland3...wayland
 - https://github.com/lbonn/rofi/pull/130

I've [just been fixing it on my end](https://gitlab.com/samueldr/nixos-configuration/-/blob/877d85b3374f8357813cc0fa3f93a61e702c007a/configuration/use-case/niri-session/default.nix#L275).

Unsure if the issue is related with the exact observations I had.

**Comment by sommerper** [comment]
I just got exactly the same error on Fedora with Niri.
```
(process:61516): libnkutils-bindings-CRITICAL **: 18:29:55.989: nk_bindings_seat_handle_key: assertion 'self->keymap != NULL' failed
```

Works fine in Hyprland. Maybe this is a Niri problem? Although it was working fine until my last reboot. Could be related to a system update?

**Reply by samueldr** [reply]
Nope, definitely ***the wayland fork*** of rofi holding the `wl_keyboard` protocol slightly wrong.

Though it does not mean it breaks on every compositor.

**Reply by JohnDowson** [reply]
It, in fact, does not break under `sway`.
Discovered this issue in the process of migrating from sway to niri.

**Reply by samueldr** [reply]
¯\\\_(ツ)\_/¯

What can I say? It's a bit more nuanced than simply it being outright crashy or not. It depends on a lot.

So the `wl_keyboard` made it so that *by the spec* the `mmap` call must be mapped `MAP_PRIVATE`.

 - https://gitlab.freedesktop.org/wayland/wayland/-/commit/905c0a341ddf0a885811d19e2b79c65a3f1d210c

In turn, `smithay`, the backend for `niri` implemented sealing

 - https://github.com/Smithay/smithay/pull/601/files

But, AFAICT, though I can't tell much since I don't know how `wlroots` really handles that in detail, `wlroots` does not have special handling for protocol version 7.

 - https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/6f6268988be93c4564e7b5be8cd46e62cd18e107/types/wlr_keyboard.c#L189-273

So if my guesstimate that `wlroots` doesn't do anything special according to the updated protocol, it would explain why it's not failing.

Note that, AFAIUI, `rofi`**`-wayland`** would need to expose itself as using protocol version 7, which will happen transitively through the [`wayland-protocols`](https://github.com/lbonn/rofi/blob/0abd88784ba5b43c44ccb9e90e8ae0262862d47b/meson.build#L96) from the system. Though note that version 7 is "new" from 1.17, so already the fork *should* be built with this new consideration in mind.

So from this surface-level look, it seems that out of two compositor backends, only one is implementing the new interface. So I would presume that in practice, to fail, the backend needs to be backed by `smithay`, or be a non-wlroots implementation that also implements the new sealed interface.

~~And what about “`hyprland`”? I don't care. They can do whatever they do, they don't deserve any attention. If something is broken in “`hyprland`”, consider using something good instead.~~

**Comment by hallettj** [comment]
This issue has been fixed in the NixOS package by applying a patch to `1.7.5+wayland3`: https://github.com/NixOS/nixpkgs/pull/356106
